src/Distribution/Audit.hs

```diff
@@ -117,14 +119,23 @@ auditMain = do
       interp = runM . interpPretty
   do
     interp do
-      buildAdvisories auditConfig nixStyleFlags
-        >>= handleBuiltAdvisories (outputHandle auditConfig) (outputFormat auditConfig)
-    `catch` \(SomeException ex) -> runM $ interpPretty do
-      owo
-        [ ([red, bold], "cabal-audit failed:\n")
-        , ([red], T.pack $ displayException ex)
-        ]
-      liftIO exitFailure
+      advisories <- buildAdvisories auditConfig nixStyleFlags
+      handleBuiltAdvisories (outputHandle auditConfig) (outputFormat auditConfig) advisories
+      when (auditConfig.failOnWarning && not (null advisories)) $ do
+        owo
+          [([red], T.pack (show (length advisories)) <> T.pack " advisories found.")]
+        liftIO exitFailure
+    `catch` \ex -> do
+      case fromException ex :: Maybe ExitCode of
+        Just (ExitFailure _code) -> exitFailure
+        Just e -> exitWith e
+        Nothing -> pure ()
+      runM $ interpPretty do
```

Comment:
I think I don't understand this block, why do we not just do the disambiguation on whether to fail or not at the very end underneath the `catch`?

Reply:
It doesn't look or read that great and feels a bit redundant. The new version looks better.